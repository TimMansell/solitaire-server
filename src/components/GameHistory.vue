<template>
  <div id="game-history">
    <p class="game-history__message">
      Showing results for all {{ userStats.completed }} games played
    </p>

    <Select
      v-model="limit"
      label="Display games per page"
      :items="['25', '50', '100', '500']"
      @select="formatLimit"
    />

    <ResponsiveTable
      :headings="['#', 'Date', 'Time', 'Outcome', 'Moves', 'Duration']"
      :items="games"
    />

    <Pagination :pages="pages" :start-on="page" @changepage="displayPage" />
  </div>
</template>

<script>
import { format, parseISO } from 'date-fns';
import numeral from 'numeral';
import VueScrollTo from 'vue-scrollto';
import { mapGetters, mapActions } from 'vuex';
import Select from '@/components/Select.vue';
import ResponsiveTable from '@/components/ResponsiveTable.vue';
import Pagination from '@/components/Pagination.vue';

const gameOutcome = ({ won, lost }) => {
  if (won) {
    return 'Won';
  }

  if (lost) {
    return 'Lost';
  }

  return 'Gave Up';
};

export default {
  name: 'GameHistory',
  components: {
    Select,
    ResponsiveTable,
    Pagination,
  },
  data() {
    return {
      page: 1,
      offset: 0,
      limit: 25,
    };
  },
  watch: {
    limit(newLimit) {
      this.offset = 0;
      this.page = 1;

      this.getAllGames({ offset: 0, limit: newLimit });
    },
  },
  computed: {
    ...mapGetters(['gameHistory', 'userStats']),
    games() {
      const {
        gameHistory,
        offset,
        userStats: { completed },
      } = this;

      const formattedGames = gameHistory.map(
        ({ won, lost, date, moves, time }, index) => ({
          number: completed - offset - index,
          date: format(parseISO(date), 'dd-MM-yyyy'),
          timePlayed: format(parseISO(date), 'HH:mm:ss'),
          outcome: gameOutcome({ won, lost }),
          moves,
          time: numeral(time).format('00:00:00'),
        })
      );

      return formattedGames;
    },
    pages() {
      const {
        limit,
        userStats: { completed },
      } = this;

      const pages = Math.ceil(completed / limit);

      return pages;
    },
  },
  async mounted() {
    const { offset, limit } = this;

    await this.getAllGames({ offset, limit });
  },
  methods: {
    ...mapActions(['getAllGames']),
    async displayPage(page) {
      const { limit } = this;
      const offset = (page - 1) * limit;

      this.offset = offset;
      this.page = page;

      await this.getAllGames({ offset, limit });

      VueScrollTo.scrollTo('#game-history', {
        container: '#history-overlay',
        offset: -10,
      });
    },
    formatLimit(limit) {
      this.limit = parseInt(limit, 10);
    },
  },
};
</script>

<style scoped lang="scss">
.game-history {
  &__message {
    padding-left: 10%;
    padding-right: 10%;
  }
}
</style>
