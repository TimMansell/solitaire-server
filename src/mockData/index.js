export const mockUid = 'f5c6a829-f0da-4dfc-81a0-e6419f0163c7';

export const mockPlayerName = 'Player Name';

export const mockHistoryApi = [
  {
    date: '2021-05-20T23:34:49.564Z',
    won: false,
    lost: false,
    moves: 1,
    time: 12,
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    won: true,
    lost: false,
    moves: 2,
    time: 12,
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    won: false,
    lost: true,
    moves: 2,
    time: 12,
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    won: true,
    lost: false,
    moves: 2,
    time: 12,
  },
];

export const mockHistory = [
  {
    date: '20-05-2021',
    time: '23:34:49',
    duration: '0:00:12',
    moves: 1,
    number: '4',
    outcome: 'Gave Up',
  },
  {
    date: '19-05-2021',
    time: '23:34:49',
    duration: '0:00:12',
    moves: 2,
    number: '3',
    outcome: 'Won',
  },
  {
    date: '19-05-2021',
    time: '23:34:49',
    duration: '0:00:12',
    moves: 2,
    number: '2',
    outcome: 'Lost',
  },
  {
    date: '19-05-2021',
    time: '23:34:49',
    duration: '0:00:12',
    moves: 2,
    number: '1',
    outcome: 'Won',
  },
];

export const mockLeaderboardsMovesAPI = [
  {
    date: '2021-04-29T12:25:47.907Z',
    uid: '7dac9d78-353f-409b-8a7f-2192409c44a2',
    moves: 2,
  },
  {
    date: '2021-04-29T12:26:20.825Z',
    uid: '2cbf658a-3102-4e9d-b749-bac853efed0d',
    moves: 2,
  },
];

export const mockLeaderboardsMoves = [
  {
    rank: 1,
    date: '29-04-2021',
    player: 'Player 1',
    moves: 2,
  },
  {
    rank: 2,
    date: '29-04-2021',
    player: 'Player 2',
    moves: 2,
  },
];

export const mockLeaderboardsTimesAPI = [
  {
    date: '2021-04-29T12:25:47.907Z',
    uid: '7dac9d78-353f-409b-8a7f-2192409c44a2',
    time: 20,
  },
  {
    date: '2021-04-29T12:26:20.825Z',
    uid: '2cbf658a-3102-4e9d-b749-bac853efed0d',
    time: 200,
  },
];

export const mockLeaderboardsTimes = [
  {
    rank: 1,
    date: '29-04-2021',
    player: 'Player 1',
    duration: '0:00:20',
  },
  {
    rank: 2,
    date: '29-04-2021',
    player: 'Player 2',
    duration: '0:03:20',
  },
];

export const mockPlayers = [
  { uid: '7dac9d78-353f-409b-8a7f-2192409c44a2', name: 'Player 1' },
  { uid: '2cbf658a-3102-4e9d-b749-bac853efed0d', name: 'Player 2' },
];

export const mockStats = { won: 2, lost: 4, completed: 9, players: 4 };

export const mockVersionNumber = '1.0.0';
