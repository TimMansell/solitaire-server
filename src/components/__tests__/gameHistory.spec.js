import { shallowMount } from '@vue/test-utils';
import GameHistory from '@/components/GameHistory.vue';
import { mockHistory } from '@/mockData';

const mocks = {
  $store: { dispatch: jest.fn() },
  $route: {
    params: {
      page: 1,
      limit: 25,
    },
  },
};

const mockComputed = {
  gameHistory: () => mockHistory,
  userGameCount: () => mockHistory.length,
};

describe('GameHistory.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should show correct completed games message', () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    expect(
      wrapper.find('[data-test="game-history-total-games"]').text()
    ).toContain('You have played a total of 4 games');
  });

  it('should show correct pages', async () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    await wrapper.setData({ filters: { limit: 3 } });

    expect(wrapper.find('[data-test="game-history-pages"]').text()).toContain(
      'Page: 1 / 2'
    );
  });

  it('should show correct showing games message ', async () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    await wrapper.setData({ filters: { limit: 3 } });

    expect(
      wrapper
        .find('[data-test="game-history-showing-games"]')
        .text()
        .replace(/\s+/g, ' ')
    ).toContain(`Showing games ${mockHistory.length} to 2`);
  });

  it('should show correct showing games message if on last page', async () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    await wrapper.setData({ filters: { limit: mockHistory.length } });

    expect(
      wrapper
        .find('[data-test="game-history-showing-games"]')
        .text()
        .replace(/\s+/g, ' ')
    ).toContain(`Showing games ${mockHistory.length} to 1`);
  });

  it('should show correct placeholder rows', async () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    await wrapper.setData({ filters: { limit: 2 } });

    expect(wrapper.vm.pageRows).toBe(2);
  });

  it('should show correct placeholder rows if rows are the same as limit', async () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    await wrapper.setData({ filters: { limit: mockHistory.length } });

    expect(wrapper.vm.pageRows).toBe(mockHistory.length);
  });

  it('should show correct placeholder rows for last page', async () => {
    const wrapper = shallowMount(GameHistory, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    await wrapper.setData({ filters: { limit: 3, page: 2 } });

    expect(wrapper.vm.pageRows).toBe(1);
  });
});
