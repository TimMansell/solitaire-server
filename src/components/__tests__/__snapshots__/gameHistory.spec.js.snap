// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameHistory.vue matches snapshot 1`] = `
<div data-test="game-history" class="game-history">
  <p data-test="game-history-total-games" data-games="4">
    You have played a total of 4 games
  </p>
  <div id="game-history-controls" data-test="game-history-controls" class="game-history__controls">
    <div data-test="game-history-pages">
      Page: 1 / 1
    </div>
    <select-stub label="Games per page" value="25" items="25,50,100,500"></select-stub>
  </div>
  <p data-test="game-history-showing-games">
    Showing games 1 to 10
  </p>
  <responsivetable-stub headings="Game,Date,Time,Outcome,Moves,Duration" items="" placeholderrows="25"></responsivetable-stub>
  <pagination-stub pages="1" starton="1"></pagination-stub>
</div>
`;
